### PR TITLE
RavenDB-10845 implementation (v4.1)

### PR DESCRIPTION
**Pull request by myarichuk**
Adding flag to DocumentStore to disable/enable compression usage by the client.
note: if any requests are in progress at the moment of changing the flag, the requests will be canceled.

### CODE REVIEW

File: src/Raven.Client/Documents/DocumentStore.cs [hunk 1]
```diff
@@ -158,28 +163,79 @@ public override RequestExecutor GetRequestExecutor(string database = null)
...
+                lazy = _requestExecutors.GetOrAdd(database, lazy);
+            }
 
             return lazy.Value;
         }
 
+        private bool _useCompression;
+
+        /// <inheritdoc />
+        public override bool UseCompression
```
Comment (ppekrol) [inline src/Raven.Client/Documents/DocumentStore.cs]: Move to Conventions? Those are frozen after DocumentStore.Initialize and cannot be changed. No need to lock anything then.

Comment (ppekrol) [reply src/Raven.Client/Documents/DocumentStore.cs]: After that most of the code will be simplified.

Comment (myarichuk) [reply src/Raven.Client/Documents/DocumentStore.cs]: Moving it to conventions will simplify code indeed. Will do so.

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 2]
```diff
@@ -126,8 +124,11 @@ private void OnFailedRequest(string url, Exception e)
         {
             FailedRequest?.Invoke(url, e);
         }
+        
+        private HttpClient GetCachedOrCreateHttpClient() => 
+            GlobalHttpClient.GetOrAdd(Certificate?.Thumbprint ?? string.Empty, new Lazy<HttpClient>(CreateClient)).Value;
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs]: You are creating this globally, so another document store with compression false will use compression

File: src/Raven.Client/Documents/Conventions/DocumentConventions.cs [hunk 3]
```diff
@@ -164,6 +164,7 @@ public DocumentConventions()
         private Func<Type, BlittableJsonReaderObject, object> _deserializeEntityFromBlittable;
         private bool _preserveDocumentPropertiesNotFoundOnModel;
         private Size _maxHttpCacheSize;
+        private bool _useCompression = true;
```
Comment (ayende) [inline src/Raven.Client/Documents/Conventions/DocumentConventions.cs]: Should be null by default here, and we will assume that null is true.

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 4]
```diff
@@ -1145,9 +1147,21 @@ public virtual void Dispose()
...
             var httpMessageHandler = new HttpClientHandler();
+            if (httpMessageHandler.SupportsAutomaticDecompression)
+            {
+                httpMessageHandler.AutomaticDecompression = 
+                    useCompression ? 
+                        DecompressionMethods.GZip | DecompressionMethods.Deflate
+                        : DecompressionMethods.None;
+            }
+            else if (httpMessageHandler.SupportsAutomaticDecompression == false &&
+                     useCompression)
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs]: Should throw only if the user explicitly asked for compression

**Review by ayende — CHANGES_REQUESTED** [review]

**Pull request merged**